pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java

```diff
@@ -0,0 +1,358 @@
+import com.linkedin.pinot.common.utils.KafkaStarterUtils;
+import com.linkedin.pinot.common.utils.ZkStarter;
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import kafka.api.FetchRequestBuilder;
+import kafka.javaapi.FetchResponse;
+import kafka.javaapi.TopicMetadata;
+import kafka.javaapi.TopicMetadataRequest;
+import kafka.javaapi.TopicMetadataResponse;
+import kafka.javaapi.consumer.SimpleConsumer;
+import kafka.javaapi.message.ByteBufferMessageSet;
+import kafka.message.MessageAndOffset;
+import org.apache.kafka.common.protocol.Errors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * TODO Document me!
+ */
+public class SimpleConsumerWrapper {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SimpleConsumerWrapper.class);
+
+  private enum ConsumerState {
+    CONNECTING_TO_BOOTSTRAP_NODE,
+    CONNECTED_TO_BOOTSTRAP_NODE,
+    FETCHING_LEADER_INFORMATION,
+    CONNECTING_TO_PARTITION_LEADER,
+    CONNECTED_TO_PARTITION_LEADER
+  }
+
+  private State _currentState;
+
+  private final boolean _metadataOnlyConsumer;
+  private final String _topic;
+  private final int _partition;
+  private final String _bootstrapNodes;
+  private SimpleConsumer _simpleConsumer;
+
+  private SimpleConsumerWrapper(String bootstrapNodes) {
+    _bootstrapNodes = bootstrapNodes;
+    _metadataOnlyConsumer = true;
+    _simpleConsumer = null;
+
+    // Topic and partition are ignored for metadata-only consumers
+    _topic = null;
+    _partition = Integer.MIN_VALUE;
+
+    setCurrentState(new ConnectingToBootstrapNode());
+  }
+
+  private SimpleConsumerWrapper(String bootstrapNodes, String topic, int partition) {
+    _bootstrapNodes = bootstrapNodes;
+    _topic = topic;
+    _partition = partition;
+    _metadataOnlyConsumer = false;
+    _simpleConsumer = null;
+
+    setCurrentState(new ConnectingToBootstrapNode());
+  }
+
+  private abstract class State {
+    private ConsumerState stateValue;
+
+    protected State(ConsumerState stateValue) {
+      this.stateValue = stateValue;
+    }
+
+    abstract void think();
+
+    abstract boolean isConnectedToKafkaBroker();
+
+    void handleConsumerException(Exception e) {
+      // TODO Remove this
+      System.out.println("e = " + e);
+      e.printStackTrace();
+
+      // By default, just log the exception and switch back to CONNECTING_TO_BOOTSTRAP_NODE (which will take care of
+      // closing the connection if it exists)
+      LOGGER.warn("Caught Kafka consumer exception, disconnecting and trying again", e);
+
+      setCurrentState(new ConnectingToBootstrapNode());
+    }
+
+    ConsumerState getStateValue() {
+      return stateValue;
+    }
+  }
+
+  private class ConnectingToBootstrapNode extends State {
+    public ConnectingToBootstrapNode() {
+      super(ConsumerState.CONNECTING_TO_BOOTSTRAP_NODE);
+    }
+
+    @Override
+    public void think() {
```

Comment:
should these be synchronized on SImpleConsumerWrapper? A single wrapper can be only in one state, right?